evm/vm/forks/sharding/log_handler.py

```diff
@@ -0,0 +1,117 @@
+import logging
+
+from eth_utils import (
+    to_dict,
+)
+
+
+def get_recent_block_hashes(w3, history_size):
+    block = w3.eth.getBlock('latest')
+    recent_hashes = [block['hash']]
+    # initialize the list of recent hashes
+    for _ in range(history_size - 1):
+        # break the loop if we hit the genesis block.
+        if block['number'] == 0:
+            break
+        block = w3.eth.getBlock(block['parentHash'])
+        recent_hashes.append(block['hash'])
+    reversed_recent_hashes = tuple(reversed(recent_hashes))
+    return reversed_recent_hashes
+
+
+def check_chain_head(w3, recent_block_hashes, history_size):
+    block = w3.eth.getBlock('latest')
+
+    new_block_hashes = []
+
+    for _ in range(history_size):
+        if block['hash'] in recent_block_hashes:
+            break
+        new_block_hashes.append(block['hash'])
+        block = w3.eth.getBlock(block['parentHash'])
+    else:
+        raise Exception('No common ancestor found for block: {0}'.format(block['hash']))
+
+    first_common_ancestor_idx = recent_block_hashes.index(block['hash'])
+
+    revoked_hashes = recent_block_hashes[first_common_ancestor_idx + 1:]
+
+    # reverse it to comply with the order of `self.recent_block_hashes`
+    reversed_new_block_hashes = tuple(reversed(new_block_hashes))
+
+    return revoked_hashes, reversed_new_block_hashes
+
+
+def preprocess_block_param(w3, block_param):
+    if isinstance(block_param, int):
+        return block_param
+    elif not isinstance(block_param, str):
+        raise ValueError("block parameter {0} must be int or str type".format(block_param))
+    current_block_number = w3.eth.blockNumber
+    if block_param == 'earliest':
+        return 0
+    elif block_param == 'latest':
+        return current_block_number
+    elif block_param == 'pending':
+        return current_block_number + 1
+    else:
+        raise ValueError(
+            "block parameter {0} in string must be one of earliest/latest/pending".format(
+                block_param,
+            )
+        )
+
+
+class LogHandler:
+
+    logger = logging.getLogger("evm.chain.sharding.LogHandler")
+
+    def __init__(self, w3, history_size=256):
+        self.history_size = history_size
+        self.w3 = w3
+        # ----------> higher score
+        self.recent_block_hashes = get_recent_block_hashes(w3, history_size)
+
+    @to_dict
+    def mk_filter_params(self, from_block_number, to_block_number, address=None, topics=None):
+        yield 'fromBlock', preprocess_block_param(self.w3, from_block_number)
+        yield 'toBlock', preprocess_block_param(self.w3, to_block_number)
+        if address is not None:
+            yield 'address', address
+        if topics is not None:
+            yield 'topics', topics
+
+    def get_new_logs(self, address=None, topics=None):
+        revoked_hashes, new_block_hashes = check_chain_head(
```

Comment:
I may have already asked this question.  Does something need to be done with the revoked hashes as well?  Should the tracker be doing anything with these to revert anything in the event that one of it's log entries for which is updated it's state is revoked?

Reply:
Actually I'm not sure about this. Please let me think about it. Sorry about this.
IMO now, If we just want to mimic the behavior of the version using `filter.get_new_entries`, I think we don't need to do the revoke.
Do you have any suggestion?

Reply:
Unfortunately, I'm not familiar enough with the underlying algorithm to be able to provide a suggestion on what the *right* thing to do here is.  Just noticed that the revoked data wasn't used and wanted be sure that wasn't an oversight.  Feel free to mark it with a `TODO` comment and open an issue to track it and we can merge with the understanding that we'll need to address it down the road.

Reply:
That's fine. Thank you so much for keeping update with this part.
Yeah you're right it is worth keeping notice of this. I marked it with `TODO` and opened the issue https://github.com/ethereum/py-evm/issues/344.